### ISSUE TITLE
V8: Copy content between languages while editing side-by-side in split view

### ISSUE THREAD

**Issue by kjac** [umbraco/Umbraco-CMS#5551]
Hi @nielslyngsoe,

I've got a PoC going for cloning content when working in split view, so the editors can quickly replicate the structure of one language in another:

![split-view-duplicate-wip](https://user-images.githubusercontent.com/7405322/58632386-99be0680-82e5-11e9-8970-54d09c94cc90.gif)

Is this meaningful? Obviously there are a few things that need sorting out still, like confirmation dialog, icon placement and of course testing a whole lot.

As for icon placement, the placement in the gif works really badly when there's a scrollbar on either side of the split view. I played around with adding the clone button to the header instead, but it also looks a bit off:

![image](https://user-images.githubusercontent.com/7405322/58632516-fa4d4380-82e5-11e9-9a0e-6d545dc797fc.png)

**Comment by kjac** [comment]
Hmmm... could also be placed between the two headers:

![image](https://user-images.githubusercontent.com/7405322/58638552-6edbae80-82f5-11e9-9c56-3e1d0cc88a33.png)

**Comment by nielslyngsoe** [comment]
Hi @kjac 

Interesting scenario, to help your feature request/contribution on its way, can I then ask you to describe a cases for this feature?

My concern is that the feature is too specific, or in other words too drastic or has too big of an impact and therefor only solves a few scenarios, by begin too specific. I would like to challenge this a little cause I see that we could end up having several interpretations of similar features (copy all properties, only paste on empty properties, replace all properties, only copy specific properties, ...)

Would it make sense to make this specific feature a bit more scenario specific?
I guess you only want to copy all properties in a situation where the language-variant is completely new. So maybe we could simplify Umbraco by only having this option available when the language-variant is empty/new...

Another though that comes to mind is if the langugage-variant already has content I think its more likely that the editor only wants one or a limited set of properties copied over, meaning this feature is properly not the go-to feature for this scenario, but should we then provide a more advanced dialog or could we look into the options for a copy feature working on property-level. As I see that concept you would be able to copy a whole property. Not necessary in split view, but in general the ability to copy a property and paste it else where. I would love to hear if you think that could solve your case or wether you think that would be help full in general.

Please enlighten me with your case and let me know what you think about these thoughts of mine.

As always, thanks for your inputs :-)

**Comment by kjac** [comment]
Hi @nielslyngsoe,

The current use case is indeed creation of a new language variant, when one of the other ones have content. 

As you point out there's a valid point in being able copy property values between language variants in general, or simply from content to content. That however is a daunting task to undertake, not least when you take permissions into account.

I guess a good option could be to spawn a dialog upon creation of a new language variant, if one or more of the variants have already been created... something along the lines of "Would you like to start from scratch or would you like to copy the contents from one of the following languages?".

Thoughts?

**Comment by nielslyngsoe** [comment]
@kjac At first I was thinking something along those lines... But I do found it a bit too drastic as a dialog, cause it would add an extra click for editors that just want to create content from scratch.

So after thinking about several possibilities I realise that I need you to share a few more details around the case for this feature.

Cause I currently only see a few specific scenarios where it would be use full, which makes it hard to find the right placement/priority for such a feature.

I see the need for this in terms of more advanced properties like the Grid or Nested Content. But simple inputs would either be variant or invariant, and therefor I guess its not in relation to these types of inputs that would need a copy, because it has already been chosen wether the field should be varying or not...

Therefor I believe this is needed for more advanced properties, is that correct?

If so, I guess the feature is needed for properties that deals with content of the page, like using Nested Content as page-builder or the grid. This scenario I can see, but the question is then, wouldn't it make more sense to just copy that specific property. That focus would simplify the feature and therefore make it easier to add to the UI.

So I really need you to describe the case. What is the context of the content, what properties are available on your content and why would the editor be interested in copying the content over from another language?

In relation to single property copy feature, you mention permissions. Can you elaborate on that, I'm not sure how the permissions comes into play when copying a selected property versus copying all properties?

Sorry for the many questions(3) and thanks for your thoughts.

I would also be open for a call — if you think it would be easier to describe the case in that way.

Thanks

**Comment by kjac** [comment]
@nielslyngsoe forget about the permissions, that was a brain fart from my side... not sure what I was thinking there 🤔 

The scenario is pretty simple, really:

1. The editors create a page in Danish and spend a looot of time in NC, essentially using it as a page builder.
2. The editors then translates the page to English. It almost always has to be an exact structural copy, just in English... so they're kinda annoyed that they have to recreate everything from scratch.

In other words, property level copying of Nested Content would likely solve 95% of this scenario. Given how that's way less intrusive to implement, I'd say that would be a good way to move forward. I'd be happy to implement a PoC if you think it makes sense?

**Comment by ronaldbarendse** [comment]
@kjac This would be a great feature and your example with NC (and with similar editors) makes a lot of sense. It might be a little specific, so I agree with @nielslyngsoe that the added logic/UI should be justified.

So I would like to make a case for creating a more generic feature, e.g. property actions. Just like with the `MenuRendering` event, you should be able to add actions based on the context of the property (content node, editor alias, current value, etc.). If the action needs additional input (like selecting the language to copy from or asking whether to override an existing value) that could be handled in a dialog, just like tree actions.

Maybe the PoC could be created as a Content App, listing the properties per group that have actions (and hide all-together when there are no actions). Content Apps aren't available in split view though (but only showing this functionality in spit view also hides it to editors not using this mode).

**Comment by kjac** [comment]
@ronaldbarendse I think you need to sketch that up to make it clearer 😄 

I have created a PoC for copying all items from one NC to another here: #5599 - input would be very welcome.

**Comment by FransdeJong** [comment]
This feature would also be helpful for grid layout, would't it? Basically for every editor that involves time to setup.

**Comment by kjac** [comment]
Definitely. NC is simply the PoC because it already incorporates copy/paste on item level.

**Comment by FransdeJong** [comment]
This could also use the blueprinting engine. Something like copy from language or pick a blueprint?
In case you want a base layout?

**Comment by kjac** [comment]
After a bunch of talks and whatnot, the conclusion is as follows: We need to provide a generic mechanism for handling copying on a property level, and implement that on Nested Content for starters. #5706 has been created for that.

**Comment by AndersBrohus-MySupport** [comment]
This would be awesome 👍

**Comment by Shazwazza** [comment]
Hi all, I am re-opening this issue since we will be refactoring it a little on how the implementation works to avoid some breaking changes introduced. 

I have reverted the changes in the 8.2 branches and have re-ported all of these changes to a new branch: https://github.com/umbraco/Umbraco-CMS/tree/v8/feature/AB3001-copy-paste-property

From there we will make some changes and get this feature introduced in 8.3. The general idea is to not have to make any c# changes and since the JS part of a property editor needs to implement the copy/paste logic anyways, we are looking for a way to ask the JS controller of the property editor if it contains a method that can be called externally. Currently this code is called by an event but I think we can potentially use the $scope that is created in the property editor controller. I haven't fully investigated this yet and it's no rush but that's the general idea. Unfortunately property editors are not true components so don't expose an API so using the $scope is sort of the poor mans version of doing this.

**Comment by kjac** [comment]
I wholeheartedly agree. And I would be more than happy to work on this; shouldn't be a huge task to port my original PR to a new solution with NC as the initial property utilizing said solution.

However. First and foremost we need to figure out how to represent "copy this editor value" in the UI as part of the editor itself (instead of the added icon next to the property label). This representation should be reused across all the editors that will eventually support copying on property level (NC, Grid, maybe MNTP, ...), so it must be fairly generic and recognizable throughout the UI.

**Comment by kjac** [comment]
@Shazwazza to get the ball rolling I had another shot at this feature; PR in #6593

**Comment by nielslyngsoe** [comment]
Progress on this topic will continue in this issue: https://github.com/umbraco/Umbraco-CMS/issues/6816

So will close this issue for now.

**Comment by kistva** [comment]
Hi @nielslyngsoe and @nul800sebastiaan 
Are there any of this issue implemented?

The idea is really good 
but I have been tracing a final solution without luck.
